feat: Force user annotation order

teddy-f-47 opened the pull request:
This PR is concerned with 2 things: (1) refactoring the store so that user state is separated from auth, (2) implementing the forcing of annotation order. Now, only the `user` store module is using the local storage (instead of all store modules), and it seems to work as intended on my machine, but please check on your environment too if auth data still gets stored into local storage. Meanwhile, the forcing of annotation order also takes advantage of the `user` store module with the `currentlyAllowedProjectId` state, and it does not affect the project admin.

What's new:
 - frontend/store/user.js : The store module that keeps data related to resting state and the project id that is allowed to be opened

What's changed:
 - frontend/store/auth.js : Refactored, cleaned back to its original code
 - frontend/plugins/persistedState.client.js : Added `paths` so that it only puts `user` in local storage
 - frontend/nuxt.config.js : Re-enabled `persistedState.client.js` plugin
 - frontend/middleware/check-auth.js : Added initialization for `user` store module
 - frontend/middleware/check-project.js : Added check on currently allowed project id to prevent user opening other projects and not the currently enforced one
 - frontend/pages/projects/_id/affective-annotation/index.vue : Refactored to use `user` store module instead of `auth`
 - frontend/components/project/ProjectList.vue : Added mechanism to force annotation order by enabling only the first project in the first page
 - frontend/pages/projects/index.vue : Added mechanism to store the project id that can be opened currently

Screenshot:
<img width="960" alt="sszxcv" src="https://user-images.githubusercontent.com/64476430/201095812-1751b327-f37b-4721-932d-304474b5ac5e.PNG">

artikandri commented on frontend/components/project/ProjectList.vue:
```diff
@@ -131,6 +136,22 @@ export default Vue.extend({
+    ...mapGetters('userState', ['getCurrentlyAllowedProjectId']),
+
+    enableProjectLink(projectId: Number) {
+      if (this.showSelect) {
```
I think rather than using this value, its better to use `isStaff` from `mapGetters` because the `showSelect` value may change depending on the requirement, what do you think?

teddy-f-47 replied:
noted 😅 I did this because I saw that showSelect is set on the project index page with the isStaff value and I thought it would be a bit redundant to add another getter, but I agree that it is better to use isStaff separately XD

artikandri commented on frontend/store/userState.js:
```diff
@@ -0,0 +1,70 @@
+}
+
+export const actions = {
+  allowProjectId({ commit }, projectId) {
```
I think the name should be `setCurrentlyAllowedProjectId` for consistency

artikandri commented on frontend/pages/projects/_id/affective-annotation/index.vue:
```diff
@@ -337,13 +337,16 @@ export default {
+        this.setArticleData()
+        this.loadLabels()
+      })
+    }
```
I think user should be notified that they can't annotate this project now, wdyt?

teddy-f-47 replied:
hello, I have added the warning message:
<img width="960" alt="ssasdhgvjhdsv" src="https://user-images.githubusercontent.com/64476430/201183603-df7f6bda-d857-4686-9189-9b89f1cb47fa.PNG">

teddy-f-47 replied:
also, for the time being I have re-enabled the persistedstate plugin. since now only `userState` module is saved into local storage, i hope it will not cause any conflict/issue. would you please check it? 😓

artikandri replied:
Hello, I got some comments I failed to notice previously, can you check it?

artikandri commented on frontend/components/project/ProjectList.vue:
```diff
@@ -131,6 +136,22 @@ export default Vue.extend({
+        return false
+      }
+    },
+    updateCurrentPage(newPage: number) {
```
probably setCurrentPage? XD

artikandri reviewed (commented):
Hello, thank you for the good work! Overall it seems fine, but I got some comments XD     Please check

artikandri commented:
Hello! I have tested the code and it works well. I have also seen the refactoring and it looks good, good job! My only comment (which I forgot to mention earlier, sorry) was to change the name of the state from `userState` to `user`, (or something else?) because we know that it's a state and the other store states don't include `state` in their name. 

As for the functionality, it works well, but I got some questions: 
1. Is this function to limit the user so that they have to annotate the projects in order only applies to `AffectiveAnnotation` project? Because I tried to access the project by url, and when I tried to annotate, it prevented me, indeed; but what about the other project types? 
2. I can still access the other functionalities (ex: view the dataset page), is it intended?

teddy-f-47 commented:
> Hello! I have tested the code and it works well. I have also seen the refactoring and it looks good, good job! My only comment (which I forgot to mention earlier, sorry) was to change the name of the state from `userState` to `user`, (or something else?) because we know that it's a state and the other store states don't include `state` in their name.
> 
> As for the functionality, it works well, but I got some questions:
> 
> 1. Is this function to limit the user so that they have to annotate the projects in order only applies to `AffectiveAnnotation` project? Because I tried to access the project by url, and when I tried to annotate, it prevented me, indeed; but what about the other project types?
> 2. I can still access the other functionalities (ex: view the dataset page), is it intended?

Thank you for the feedback! I will rename the state to `user`.
1. It is a very good question related to a point I forgot to mention -- currently it affects all project types in the user (annotator) account. So, if user X gets assigned multiple projects with different types (i.e. text classification, sequence to sequence, affective annotation), they still need to do the annotation in order, from first project to the last, without regards to the project type. I'm not sure if we should allow them to annotate other project types outside of affective annotation, since such thing was not mentioned in the requirements, but in my opinion it makes more sense to keep the behavior as it is right now (?)
2. Yes, it is intended. I think preventing the annotation in the annotation page alone is enough, as there is not much that they can do anyway in the dataset menu page. But I'm open to suggestion if we need to "block" them from seein the dataset page as well.

teddy-f-47 commented:
Hello, I have renamed the module from `userState` to `user`, and I have added a `clear` function which can be used to reset the state data in the local storage (i.e. set the states to `null`). I'm afraid there is no easy way to actually "delete" states variables from local storage ([link](https://github.com/robinvdvleuten/vuex-persistedstate/issues/88)) though.

artikandri commented:
1. I think so; however, when I tried accessing the article annotation project, the page still opens. The currently allowed project id is 2 and my project id is 93, but I can still access the annotation page. I think this is because you only handled the condition in `affective-annotation/index.vue`, but it should be handled in other pages as well. 
![image](https://user-images.githubusercontent.com/12537724/201647324-768fdf0e-ad70-4462-997f-ec56ed3657e5.png)

My suggestion is to use the `router middleware` to check the page url before going to the page, and redirect the user back when they cant access the page. You can also implement a new page to notify user that they cant annotate the project until they finished annotating current project.

haanh764 reviewed (commented):
Hello, thank you for your great work XD I have tested the feature to force the project order for annotating and it works fine for now (regardless the point that Argi has commented). 
However, I am having a weird behaviour with the resting time :sweat_smile: 
I have finished the annotation and as you can see in the picture below it said that I can continue at 17:04, but my time is already 17:11 and I tried F5 many times but still cant access. I also logged out and logged in again. Can you please check if it also happens in your machine or only in my machine :sweat_smile: 

Other than that, I think the users' process are separated oke I can log in as another user without no problem.
![image](https://user-images.githubusercontent.com/17596118/201710159-63106665-0ee1-4bc1-b283-9e2189bfc744.png)

teddy-f-47 commented:
> My suggestion is to use the `router middleware` to check the page url before going to the page, and redirect the user back when they cant access the page. You can also implement a new page to notify user that they cant annotate the project until they finished annotating current project.

Hello! I have moved the project id checking into the `check-project.js` middleware, as I think such a mechanism should be put inside this middleware (together with the checking if the user is opening an uncompleted project). So, I have removed the warning alert on the annotation page, as it is no longer necessary, since the user will be automatically redirected to the project list page if they try entering the project url manually. 😅

teddy-f-47 commented:
> Hello, thank you for your great work XD I have tested the feature to force the project order for annotating and it works fine for now (regardless the point that Argi has commented). However, I am having a weird behaviour with the resting time 😅 I have finished the annotation and as you can see in the picture below it said that I can continue at 17:04, but my time is already 17:11 and I tried F5 many times but still cant access. I also logged out and logged in again. Can you please check if it also happens in your machine or only in my machine 😅

Hello, thank you for finding the bug! 😅 I did find the bug as well. However, I've just added the commit for the fix. Interestingly, I actually found such a bug a long time ago during initial development of the rest break, and I have fixed it in PR 40, but somehow I changed back the code in PR 45. It is my mistake, but I'm just not sure why it happened. 😅

artikandri commented on frontend/pages/projects/index.vue:
```diff
@@ -99,13 +99,26 @@ export default Vue.extend({
+        const nextProjectIdToAnnotate = progress?.project_id || -1
+        this.setCurrentlyAllowedProjectId(nextProjectIdToAnnotate)
+      } catch {
+        console.log("error in checking project annotation order")
```
this console.log is no longer needed, no?

artikandri commented on frontend/components/project/ProjectList.vue:
```diff
@@ -87,7 +89,8 @@ export default Vue.extend({
       options: {} as DataOptions,
-      mdiMagnify
+      mdiMagnify,
+      projects: this.items
```
Is this `projects` variable used anywhere?

teddy-f-47 replied:
No, I initially used it for a different approach but now it is not used. Sorry I missed removing this line.

artikandri commented on frontend/store/user.js:
```diff
@@ -0,0 +1,79 @@
+    const restingEndTime = getters.getRestingEndTime
+
+    if (currentTime >= restingEndTime) {
+      commit('clearRestingPeriod')
```
I dont remember if I missed this previously, but I think a getter should only do one thing, to `get` the data, and not to clear the data :"   how about separating this to a new function?

teddy-f-47 replied:
Hello, this one is actually not a "getter", it is actually an action. I need the action to wrap the real getter (that one on the top with `getters.getRestingEndTime` together with mechanism to clear the resting time data when the resting period has finished. Notably, the real getter (`getters.getRestingEndTime`) only does one thing which is getting the data. Maybe I should rename this action from `getRestingPeriod` to something like `checkRestingPeriod` or `calculateRestingPeriod`?

artikandri replied:
Yes, I think that will be more clear XD sorry for the long reply

artikandri replied:
Or probably you can separate xD so we can have both functions, get... and clear... Depending on your preference.

artikandri commented on frontend/pages/projects/index.vue:
```diff
@@ -99,13 +99,26 @@ export default Vue.extend({
 
+    findNextProjectIdToAnnotate(progressList: MyProgressList) {
+      try {
+        const progresses = _.cloneDeep(progressList)
```
I don't think cloneDeep to `progresses` variable is needed here, wdyt? Just use the `progressList` directly

artikandri reviewed (commented):
Hello, thank you for the good work! I have tested the parts that I commented before (aside from Anh's comment) and it works well, nice job! However, I still have some things to comment about, sorry :")    Can you check?

haanh764 reviewed (approved):
Hello, thank you for your quick fix XD I have tested and everything works fine for me. Approved for merging XD

artikandri reviewed (approved):
Hello, I have checked the code and aside from the getter, I think it looks oke now xD  sorry for the many comments and thank you for fixing them. Approved for merging!

teddy-f-47 commented:
Hello, I have renamed the function from `getRestingPeriod` to `calculateRestingPeriod`. I also did a minor refactor inside the function to make it look neater and easier to understand. I will merge this PR now. XD Thank you!
